rfd/0038-resource-encryption.md

```diff
@@ -0,0 +1,91 @@
+
+Option 2 here seems like a far more robust solution for production use and that is thus what we should employ.
+
+### Encryption format
+
+For facilitating key rotation the encrypted objects should be stored in a seperate file from it's data key bundle. This is due to modification restricts on objects imposed by services such as AWS S3.
+
+All encrypted objects should be suffixed by a `.enc` extension to communicate that they are encrypted. For simplicity no other data should be bundled in the same file.
+
+The file that stores the encrypted data key should be a file with the name `$objectName.key` stored next to the encrypted object.
+
+Each encrypted object has an accompanying file named `$objectName.key` that stores a JSON document of this format.
+
+```json
+{
+    // Contains a base64 encoded AES-CBC-HMACSHA256 data key encrypted with the master key.
+    "dataKey": "string",
+
+    // An integer starting at 1 one that communicates the schema used for encryption.
+    // May be changed in the future.
+    "version": "number",
+
+    // A hex encoded version of the SHA-256 hash of the master key.
+    // This is used to determine state during key rotation.
+    "masterKey": "string",
+}
+```
+
+This format allows a decrypting auth server to easily fetch and decrypt the data key using the master key. It can then fetch and decrypt the object itself.
+
+Key rotation in event of a security incident can be performed by fetching and rewriting every key file to reencrypt the data key.
+
+The format detailed above should be used wherever it makes sense from a security standpoint. Devitations from how the data object is encrypted should be made wherever necessary but the key bundle format should never be deviated from. A notable deviation that will be made is for Parquet based files where the encryption will be handled by giving Parquet control of the symmetric data key. The reason for this is that flat-file encryption with Parquet is vulnerable to security attacks similar to CRIME.
+
+### Key configuration
+
+The encryption and key management schema detailed above calls of a central store containing a PEM RSA-2048 keypair used to encrypt the data keys. This will be implemented by introducing a new sensitive resource that is only ever present on auth servers containing the key.
+
+If the resource isn't present on auth server startup, a new key is randomly generated via an OS CSPRNG and the resource is set via a compare-and-swap. If the compare-and-swap fails another auth server already created the resource and we fetch that instead.
+
+### HSM Support
+
+This system utilizes HSMs in a number of different ways. When a HSM is configured during setup or key rotation, it is used to securely generate new master keys. When supported, HSMs should also be used for encrypting and decrypting data keys using the master key.
```

Comment:
Consider that if you ever want to support HSMs in an HA cluster with multiple auth servers, each auth server will have its own HSM (so you may need to support multiple "master" keys)

Reply:
Is there no option to load the RSA master keys into the HSM? I did get that impression but I may very well be wrong here.